### PR TITLE
Add rolify gem to manage school permissions

### PR DESCRIPTION
**Pull request by ethax-ross**
### Trello card

[Trello-664](https://trello.com/c/NUHQxJWP/664-add-role-library-for-controlling-access-to-schools)

### Context

We want to manage permissions within the app going forward (instead of via DfE sign in). To do this we need a way of tracking who has access to which resource/school.

### Changes proposed in this pull request

- Add rolify gem to manage roles

Add `rolify` gem; this is a simple and well maintained gem that will enable us to manage user roles on a per-resource/school basis.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class User < ApplicationRecord
+  rolify
```
Comment (asatwal) [inline app/models/user.rb:2]: I know this is part of the last PR, but does `dfe_sign_in_user` have to be a `attr_writer` ? Could it be an `attr_reader` ?

**Review by asatwal — COMMENTED** [review]
Just a few comments. 

There are a few questions I have about Rolify:



(1) How are the roles actually managed ? Will we have some admin screens to manage Roles ?

(2) When a User is created in the App will they get a default Role ?

(3) Do we have to handle legacy Roles, whether they be explicit or implicit ?

(4) How are Roles (if any?)  currently managed, and how will these map onto the new model.

**Comment by ethax-ross** [comment]
> Just a few comments.
> 
> There are a few questions I have about Rolify:
> 
> (1) How are the roles actually managed ? Will we have some admin screens to manage Roles ?
> 
> (2) When a User is created in the App will they get a default Role ?
> 
> (3) Do we have to handle legacy Roles, whether they be explicit or implicit ?
> 
> (4) How are Roles (if any?) currently managed, and how will these map onto the new model.

1) Yeah exactly that; the next PR will add a 'manage users' section to the school dashboard where we display dfe sign in users associated with the school/organisation and they can be approved/denied access.

2) For now they aren't but the idea is when we migrate across we will backfill roles for existing users onto their new `User` model.

3 and 4) Roles are currently managed in DfE sign in; to access GSE you need to be part of the organisation/school in DfE sign in and have the 'school experience service' role/access. This is whats causing our users issues; they are often part of the organisation/school in DfE sign in but don't understand the second step of requesting the school experience service access/role. Going forward that part will be handled in the GSE app by Rolify, so you need to be part of the org/school in DfE sign in still but after that someone else at the school can approve your access in the GSE school portal.

File: app/models/user.rb (hunk 1 above)
Comment (asatwal) [reply app/models/user.rb:2]: I will talk to you on this at some point. OK for now, as this is still evolving.

**Comment by ethax-ross** [comment]
We will be using a new DfE service account for this new mechanism so there shouldn't be anything to keep in sync. If a user is revoked access from the organisation in DfE sign in then they won't be able to log in and will lose access as soon as their session expires. You will need both a valid DfE sign in account with access to the organisation/school and the relevant role in GSE.

**Pull request merged**